### ISSUE TITLE
It seems that shim doesn't support select() on /dev/urandom 

### ISSUE THREAD

**Issue by shaoyuta** [oscarlab/graphene#1319]
1, I taste to execute nodejs in pal, reference to https://github.com/oscarlab/graphene-tests/tree/master/nodejs. My operations is as below:
```
$ cd LibOS/shim/test/apps/nodejs
$ make
$ ./pal_loader nodejs.manifest helloworld.js  
```
But exception occoupied:
```
*** An unexpected memory fault occurred inside PAL. Exiting the thread. (PID = 6211, TID = 6211, RIP = +0x00003bbd) ***
```
2, Debug it and found that the exception is due to nodejs try to access file "/dev/urandom" and use the api "select()". I write a simple app to test it
```
#define TESTFILE "/dev/urandom"

int main(int argc, char** argv) {
    int fd;
    struct timeval t;
    fd_set fset;
    FD_ZERO(&fset);

    fd = open(TESTFILE, O_RDONLY  | O_NONBLOCK | O_NOCTTY);
    if( fd < 0 ){
	    printf("open file failure\n");
        return -1;
    }

    FD_SET(fd, &fset);

    t.tv_sec = 1;
    t.tv_usec = 0;
    if (select(fd + 1, &fset, NULL, NULL, &t) >= 0) {
        if (FD_ISSET(fd, &fset))
            printf("ok\n");
    }

    return 0;
}
```
The program can run normally in the native environment, but still abnormal in the pal environment.
```
$ ./pal_loader  helloworld
[P10131] host: Linux
[P10131] bkeep_unmapped: 0x7f1f914cb000-0x7f1f914cd000 => vma
[P10131] allocate 0x7f1f914cb000-0x7f1f914cd000 for vmas
[P10131] Converting early VMA [0x7f1f919d8080] 0x7f1f914cd000-0x7f1f916ce018
[P10131] Converting early VMA [0x7f1f919d80c8] 0x7f1f916cf000-0x7f1f919dc000
[P10131] Converting early VMA [0x7f1f919d8110] 0x7f1f914cb000-0x7f1f914cd000
[P10131] heap top adjusted to 0x491a8ed9f000
[P10131] bkeep_unmapped: 0x7f1f914ba000-0x7f1f914cb000 => slab
[P10131] bkeep_unmapped: 0x7f1f91499000-0x7f1f914ba000 => slab
[P10131] bkeep_unmapped: 0x7f1f91495000-0x7f1f91499000 => slab
[P10131] bkeep_unmapped: 0x7f1f91490000-0x7f1f91495000 => slab
[P10131] bkeep_unmapped: 0x7f1f91488000-0x7f1f91490000 => slab
[P10131] shim loaded at 0x7f1f916cf000, ready to initialize
[P10131] bkeep_unmapped: 0x7f1f91487000-0x7f1f91488000 => manifest
[P10131] bkeep_unmapped: 0x7f1f91486000-0x7f1f91487000 => slab
[P10131] bkeep_unmapped: 0x7f1f91484000-0x7f1f91486000 => slab
[P10131] mounting default root filesystem
[P10131] bkeep_unmapped: 0x7f1f91482000-0x7f1f91484000 => slab
[P10131] bkeep_unmapped: 0x7f1f91480000-0x7f1f91482000 => slab
[P10131] bkeep_munmap: 0x7f1f91482000-0x7f1f91484000
[P10131] bkeep_munmap: 0x7f1f91480000-0x7f1f91482000
[P10131] mounting as proc filesystem: /proc
[P10131] mounting as dev filesystem: /dev
[P10131] mounting as chroot filesystem: from dev:tty to /dev
[P10131] bkeep_unmapped: 0x7f1f91483000-0x7f1f91484000 => slab
[P10131] bkeep_unmapped: 0x7f1f91482000-0x7f1f91483000 => slab
[P10131] creating pipe: pipe.srv:625551251
[P10131] Adding port (handle 0x7f1f91bfb850) for process 10131 (type 0002)
[P10131] walk_thread_list(callback=0x7f1f91705d6e)
[P10131] allocated pid: 1
[P10131] bkeep_unmapped: 0x7f1f9147d000-0x7f1f91482000 => slab
[     1] mounting as chroot filesystem: from file:..//../../../Runtime to /lib
[     1] bkeep_unmapped: 0x7f1f9147b000-0x7f1f9147d000 => slab
[     1] bkeep_munmap: 0x7f1f9147b000-0x7f1f9147d000
[     1] bkeep_unmapped: 0x7f1f9147b000-0x7f1f9147d000 => slab
[     1] bkeep_munmap: 0x7f1f9147b000-0x7f1f9147d000
[     1] mounting as chroot filesystem: from file:/bin to /bin
[     1] bkeep_unmapped: 0x7f1f9147b000-0x7f1f9147d000 => slab
[     1] bkeep_unmapped: 0x7f1f91479000-0x7f1f9147b000 => slab
[     1] bkeep_munmap: 0x7f1f9147b000-0x7f1f9147d000
[     1] bkeep_munmap: 0x7f1f91479000-0x7f1f9147b000
[     1] bkeep_unmapped: 0x7f1f9147b000-0x7f1f9147d000 => slab
[     1] bkeep_unmapped: 0x7f1f91479000-0x7f1f9147b000 => slab
[     1] bkeep_munmap: 0x7f1f9147b000-0x7f1f9147d000
[     1] bkeep_munmap: 0x7f1f91479000-0x7f1f9147b000
[     1] bkeep_unmapped: 0x491a8e99e000-0x491a8ed9f000 => stack
[     1] heap top adjusted to 0x491a8e99e000
[     1] bkeep_mprotect: 0x491a8e99f000-0x491a8ed9f000
[     1] allocated stack at 0x491a8e99f000 (size = 4194304)
[     1] adding file:helloworld as runtime object loaded at 0x7f1f914cd000-0x7f1f916ce018
[     1] bkeep_mprotect: 0x7f1f914ce000-0x7f1f916cd000
[     1] bkeep_mmap: 0x7f1f916cd000-0x7f1f916cf000
[     1] bkeep_unmapped: 0x491a8c99e000-0x491a8e99e000 => brk
[     1] heap top adjusted to 0x491a8c99e000
[     1] brk area: 0x491a8c99e000 - 0x491a8c99f000
[     1] brk reserved area: 0x491a8c99f000 - 0x491a8e99e000
[     1] bkeep_mmap: 0x491a8c99e000-0x491a8c99f000
[     1] search interpreter: /lib/ld-linux-x86-64.so.2
[     1] bkeep_unmapped: 0x7f1f91479000-0x7f1f9147d000 => filebuf
[     1] bkeep_unmapped: 0x491a8c777000-0x491a8c99e000
[     1] heap top adjusted to 0x491a8c777000
[     1] bkeep_mprotect: 0x491a8c79c000-0x491a8c99b000
[     1] bkeep_mmap: 0x491a8c99b000-0x491a8c99d000
[     1] bkeep_mmap: 0x491a8c99d000-0x491a8c99e000
[     1] add a library for gdb: file:..//../../../Runtime/ld-linux-x86-64.so.2
[     1] shim process initialized
[     1] bkeep_unmapped: 0x491a8c776000-0x491a8c777000 => linux-vdso.so.1
[     1] heap top adjusted to 0x491a8c776000
[P10131] bkeep_unmapped: 0x7f1f91474000-0x7f1f91479000 => slab
[     1] ---- shim_brk(0x0) = 0x491a8c99e000
[P10131] bkeep_mprotect: 0x7f1f91475000-0x7f1f91479000
[P10131] allocated stack at 0x7f1f91475000 (size = 16384)
[P10131] IPC helper thread started
[     1] ---- shim_access("/etc/ld.so.preload",F_OK|R_OK) = -2
[P10131] Listening to process 10131 on port 0x7f1f91482048 (handle 0x7f1f91bfb850, type 0002)
[     1] ---- shim_openat(AT_FDCWD,"/lib/tls/haswell/x86_64/libc.so.6",O_RDONLY|2000000,0000) = -2
[     1] ---- shim_stat("/lib/tls/haswell/x86_64",0x491a8ed9d130) = -2
[     1] ---- shim_openat(AT_FDCWD,"/lib/tls/haswell/libc.so.6",O_RDONLY|2000000,0000) = -2
[     1] ---- shim_stat("/lib/tls/haswell",0x491a8ed9d130) = -2
[     1] ---- shim_openat(AT_FDCWD,"/lib/tls/x86_64/libc.so.6",O_RDONLY|2000000,0000) = -2
[     1] ---- shim_stat("/lib/tls/x86_64",0x491a8ed9d130) = -2
[     1] ---- shim_openat(AT_FDCWD,"/lib/tls/libc.so.6",O_RDONLY|2000000,0000) = -2
[     1] ---- shim_stat("/lib/tls",0x491a8ed9d130) = -2
[     1] ---- shim_openat(AT_FDCWD,"/lib/haswell/x86_64/libc.so.6",O_RDONLY|2000000,0000) = -2
[     1] ---- shim_stat("/lib/haswell/x86_64",0x491a8ed9d130) = -2
[     1] ---- shim_openat(AT_FDCWD,"/lib/haswell/libc.so.6",O_RDONLY|2000000,0000) = -2
[     1] ---- shim_stat("/lib/haswell",0x491a8ed9d130) = -2
[     1] ---- shim_openat(AT_FDCWD,"/lib/x86_64/libc.so.6",O_RDONLY|2000000,0000) = -2
[     1] ---- shim_stat("/lib/x86_64",0x491a8ed9d130) = -2
[     1] ---- shim_openat(AT_FDCWD,"/lib/libc.so.6",O_RDONLY|2000000,0000) = 3
[     1] ---- shim_read(3,0x491a8ed9d298,832) ... read
[     1] bkeep_unmapped: 0x7f1f91470000-0x7f1f91474000 => filebuf
[     1] ---- return from shim_read(...) = 832
[     1] ---- shim_fstat(3,0x491a8ed9d130) = 0
[     1] ---- shim_mmap(0x0,8192,PROT_READ|PROT_WRITE,MAP_PRIVATE|MAP_ANON,-1,0) ... mmap
[     1] bkeep_unmapped: 0x491a8c774000-0x491a8c776000
[     1] heap top adjusted to 0x491a8c774000
[     1] ---- return from shim_mmap(...) = 0x491a8c774000
[     1] ---- shim_mmap(0x0,3869368,PROT_READ|PROT_EXEC,MAP_PRIVATE|0x800,3,0) ... mmap
[     1] bkeep_unmapped: 0x491a8c3c3000-0x491a8c774000
[     1] heap top adjusted to 0x491a8c3c3000
[     1] ---- return from shim_mmap(...) = 0x491a8c3c3000
[     1] ---- shim_mprotect(0x491a8c56a000,2097152,PROT_NONE) ... mprotect
[     1] bkeep_mprotect: 0x491a8c56a000-0x491a8c76a000
[     1] ---- return from shim_mprotect(...) = 0
[     1] ---- shim_mmap(0x491a8c76a000,24576,PROT_READ|PROT_WRITE,MAP_PRIVATE|MAP_FIXED|0x800,3,1732608) ... mmap
[     1] mmap: allowing overlapping MAP_FIXED allocation at 0x491a8c76a000 with length 24576
[     1] bkeep_mmap: 0x491a8c76a000-0x491a8c770000
[     1] ---- return from shim_mmap(...) = 0x491a8c76a000
[     1] ---- shim_mmap(0x491a8c770000,15032,PROT_READ|PROT_WRITE,MAP_PRIVATE|MAP_ANON|MAP_FIXED,-1,0) ... mmap
[     1] mmap: allowing overlapping MAP_FIXED allocation at 0x491a8c770000 with length 16384
[     1] bkeep_mmap: 0x491a8c770000-0x491a8c774000
[     1] ---- return from shim_mmap(...) = 0x491a8c770000
[     1] ---- shim_close(3) = 0
[     1] glibc register library /lib/libc.so.6 loaded at 0x491a8c3c3000
[     1] add a library for gdb: file:..//../../../Runtime/libc.so.6
[     1] set fs_base to 0x491a8c775500
[     1] ---- shim_arch_prctl(4098,0x491a8c775500) = 0x00000000
[     1] ---- shim_mprotect(0x491a8c76a000,16384,PROT_READ) ... mprotect
[     1] bkeep_mprotect: 0x491a8c76a000-0x491a8c76e000
[     1] ---- return from shim_mprotect(...) = 0
[     1] ---- shim_mprotect(0x7f1f916cd000,4096,PROT_READ) ... mprotect
[     1] bkeep_mprotect: 0x7f1f916cd000-0x7f1f916ce000
[     1] ---- return from shim_mprotect(...) = 0
[     1] ---- shim_mprotect(0x491a8c99b000,4096,PROT_READ) ... mprotect
[     1] bkeep_mprotect: 0x491a8c99b000-0x491a8c99c000
[     1] ---- return from shim_mprotect(...) = 0
[     1] ---- shim_openat(AT_FDCWD,"/dev/urandom",O_RDONLY|4400,0000) = 3
*** An unexpected memory fault occurred inside PAL. Exiting the thread. (PID = 10131, TID = 10131, RIP = +0x00003bbd) ***
```

**Comment by mkow** [comment]
https://github.com/oscarlab/graphene/commit/3def3a6ba6358a489223daf18eaa141f80ee0a7a from #1253 looks related.

Our `select` implementation is actually a wrapper on `poll`, which is tested against `/dev` usage, so it's weird that it doesn't work.
Could you check if it also crashes when using `poll`?

**Comment by shaoyuta** [comment]
@mkow, Thanks for your reply.

More detail information is as below:
```
$ lsb_release -a
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 18.04.4 LTS
Release:	18.04
Codename:	bionic
```
Last commit of graphene is :  **40c08ff75ae4c1034632e165f308aa7d7839358f**


I use api poll() to test it again, the app still crash.
```
#define TESTFILE "/dev/urandom"

int main(int argc, char** argv) {
    int fd;
    struct timeval t;
    fd_set fset;
    FD_ZERO(&fset);

    fd = open(TESTFILE, O_RDONLY  | O_NONBLOCK | O_NOCTTY);
    if( fd < 0 ){
	    printf("open file failure\n");
        return -1;
    }

    struct pollfd pfds[1];
    pfds[0].fd=fd;
    pfds[0].events = POLLRDNORM | POLLIN;
    poll(pfds, 1, -1);
    printf("0x%x\n",pfds[0].revents);
    if (pfds[0].revents & POLLRDNORM) {
        printf("ok\n");
    }
    return 0;
}
```
manifest:
```
loader.preload = file:..//../../../Runtime/libsysdb.so
loader.env.LD_LIBRARY_PATH = /lib
loader.debug_type = inline
loader.syscall_symbol = syscalldb

fs.mount.lib.type = chroot
fs.mount.lib.path = /lib
fs.mount.lib.uri = file:..//../../../Runtime

fs.mount.bin.type = chroot
fs.mount.bin.path = /bin
fs.mount.bin.uri = file:/bin

sys.brk.size = 32M
sys.stack.size = 4M

# allow to bind on port 8000
net.allow_bind.1 = 127.0.0.1:8000
# allow to connect to port 8000
net.allow_peer.1 = 127.0.0.1:8000

# sgx-related
sgx.trusted_files.ld = file:..//../../../Runtime/ld-linux-x86-64.so.2
sgx.trusted_files.libc = file:..//../../../Runtime/libc.so.6
sgx.trusted_files.libdl = file:..//../../../Runtime/libdl.so.2
sgx.trusted_files.libm = file:..//../../../Runtime/libm.so.6
sgx.trusted_files.libpthread = file:..//../../../Runtime/libpthread.so.0

sgx.trusted_files.tcp_c = file:tcp.c
```

The crash log is as below:
```
[P17230] host: Linux
[P17230] bkeep_unmapped: 0x7fb057df3000-0x7fb057df5000 => vma
[P17230] allocate 0x7fb057df3000-0x7fb057df5000 for vmas
[P17230] Converting early VMA [0x7fb058300080] 0x7fb057df5000-0x7fb057ff6018
[P17230] Converting early VMA [0x7fb0583000c8] 0x7fb057ff7000-0x7fb058304000
[P17230] Converting early VMA [0x7fb058300110] 0x7fb057df3000-0x7fb057df5000
[P17230] heap top adjusted to 0x71102b0df000
[P17230] bkeep_unmapped: 0x7fb057de2000-0x7fb057df3000 => slab
[P17230] bkeep_unmapped: 0x7fb057dc1000-0x7fb057de2000 => slab
[P17230] bkeep_unmapped: 0x7fb057dbd000-0x7fb057dc1000 => slab
[P17230] bkeep_unmapped: 0x7fb057db8000-0x7fb057dbd000 => slab
[P17230] bkeep_unmapped: 0x7fb057db0000-0x7fb057db8000 => slab
[P17230] shim loaded at 0x7fb057ff7000, ready to initialize
[P17230] bkeep_unmapped: 0x7fb057daf000-0x7fb057db0000 => manifest
[P17230] bkeep_unmapped: 0x7fb057dae000-0x7fb057daf000 => slab
[P17230] bkeep_unmapped: 0x7fb057dac000-0x7fb057dae000 => slab
[P17230] mounting default root filesystem
[P17230] bkeep_unmapped: 0x7fb057daa000-0x7fb057dac000 => slab
[P17230] bkeep_unmapped: 0x7fb057da7000-0x7fb057daa000 => slab
[P17230] bkeep_unmapped: 0x7fb057da2000-0x7fb057da7000 => slab
[P17230] bkeep_munmap: 0x7fb057da7000-0x7fb057daa000
[P17230] bkeep_unmapped: 0x7fb057d9d000-0x7fb057da2000 => slab
[P17230] bkeep_munmap: 0x7fb057da2000-0x7fb057da7000
[P17230] bkeep_munmap: 0x7fb057daa000-0x7fb057dac000
[P17230] bkeep_munmap: 0x7fb057d9d000-0x7fb057da2000
[P17230] mounting as proc filesystem: /proc
[P17230] mounting as dev filesystem: /dev
[P17230] mounting as chroot filesystem: from dev:tty to /dev
[P17230] bkeep_unmapped: 0x7fb057dab000-0x7fb057dac000 => slab
[P17230] bkeep_unmapped: 0x7fb057daa000-0x7fb057dab000 => slab
[P17230] creating pipe: pipe.srv:741557070
[P17230] Adding port (handle 0x7fb058523850) for process 17230 (type 0002)
[P17230] walk_thread_list(callback=0x7fb05802dd6e)
[P17230] allocated pid: 1
[P17230] bkeep_unmapped: 0x7fb057da5000-0x7fb057daa000 => slab
[     1] mounting as chroot filesystem: from file:..//../../../Runtime to /lib
[     1] bkeep_unmapped: 0x7fb057da3000-0x7fb057da5000 => slab
[     1] bkeep_munmap: 0x7fb057da3000-0x7fb057da5000
[     1] bkeep_unmapped: 0x7fb057da3000-0x7fb057da5000 => slab
[     1] bkeep_munmap: 0x7fb057da3000-0x7fb057da5000
[     1] mounting as chroot filesystem: from file:/bin to /bin
[     1] bkeep_unmapped: 0x7fb057da3000-0x7fb057da5000 => slab
[     1] bkeep_unmapped: 0x7fb057da1000-0x7fb057da3000 => slab
[     1] bkeep_munmap: 0x7fb057da3000-0x7fb057da5000
[     1] bkeep_munmap: 0x7fb057da1000-0x7fb057da3000
[     1] bkeep_unmapped: 0x7fb057da3000-0x7fb057da5000 => slab
[     1] bkeep_unmapped: 0x7fb057da1000-0x7fb057da3000 => slab
[     1] bkeep_munmap: 0x7fb057da3000-0x7fb057da5000
[     1] bkeep_munmap: 0x7fb057da1000-0x7fb057da3000
[     1] bkeep_unmapped: 0x71102acde000-0x71102b0df000 => stack
[     1] heap top adjusted to 0x71102acde000
[     1] bkeep_mprotect: 0x71102acdf000-0x71102b0df000
[     1] allocated stack at 0x71102acdf000 (size = 4194304)
[     1] adding file:poll as runtime object loaded at 0x7fb057df5000-0x7fb057ff6018
[     1] bkeep_mprotect: 0x7fb057df6000-0x7fb057ff5000
[     1] bkeep_mmap: 0x7fb057ff5000-0x7fb057ff7000
[     1] bkeep_unmapped: 0x711028cde000-0x71102acde000 => brk
[     1] heap top adjusted to 0x711028cde000
[     1] brk area: 0x711028cde000 - 0x711028cdf000
[     1] brk reserved area: 0x711028cdf000 - 0x71102acde000
[     1] bkeep_mmap: 0x711028cde000-0x711028cdf000
[     1] search interpreter: /lib/ld-linux-x86-64.so.2
[     1] bkeep_unmapped: 0x7fb057da1000-0x7fb057da5000 => filebuf
[     1] bkeep_unmapped: 0x711028ab7000-0x711028cde000
[     1] heap top adjusted to 0x711028ab7000
[     1] bkeep_mprotect: 0x711028adc000-0x711028cdb000
[     1] bkeep_mmap: 0x711028cdb000-0x711028cdd000
[     1] bkeep_mmap: 0x711028cdd000-0x711028cde000
[     1] add a library for gdb: file:..//../../../Runtime/ld-linux-x86-64.so.2
[     1] shim process initialized
[     1] bkeep_unmapped: 0x711028ab6000-0x711028ab7000 => linux-vdso.so.1
[     1] heap top adjusted to 0x711028ab6000
[P17230] bkeep_unmapped: 0x7fb057d9c000-0x7fb057da1000 => slab
[P17230] bkeep_mprotect: 0x7fb057d9d000-0x7fb057da1000
[P17230] allocated stack at 0x7fb057d9d000 (size = 16384)
[P17230] IPC helper thread started
[     1] ---- shim_brk(0x0) = 0x711028cde000
[P17230] Listening to process 17230 on port 0x7fb057daa048 (handle 0x7fb058523850, type 0002)
[     1] ---- shim_access("/etc/ld.so.preload",F_OK|R_OK) = -2
[     1] ---- shim_openat(AT_FDCWD,"/lib/tls/haswell/x86_64/libc.so.6",O_RDONLY|2000000,0000) = -2
[     1] ---- shim_stat("/lib/tls/haswell/x86_64",0x71102b0dd120) = -2
[     1] ---- shim_openat(AT_FDCWD,"/lib/tls/haswell/libc.so.6",O_RDONLY|2000000,0000) = -2
[     1] ---- shim_stat("/lib/tls/haswell",0x71102b0dd120) = -2
[     1] ---- shim_openat(AT_FDCWD,"/lib/tls/x86_64/libc.so.6",O_RDONLY|2000000,0000) = -2
[     1] ---- shim_stat("/lib/tls/x86_64",0x71102b0dd120) = -2
[     1] ---- shim_openat(AT_FDCWD,"/lib/tls/libc.so.6",O_RDONLY|2000000,0000) = -2
[     1] ---- shim_stat("/lib/tls",0x71102b0dd120) = -2
[     1] ---- shim_openat(AT_FDCWD,"/lib/haswell/x86_64/libc.so.6",O_RDONLY|2000000,0000) = -2
[     1] ---- shim_stat("/lib/haswell/x86_64",0x71102b0dd120) = -2
[     1] ---- shim_openat(AT_FDCWD,"/lib/haswell/libc.so.6",O_RDONLY|2000000,0000) = -2
[     1] ---- shim_stat("/lib/haswell",0x71102b0dd120) = -2
[     1] ---- shim_openat(AT_FDCWD,"/lib/x86_64/libc.so.6",O_RDONLY|2000000,0000) = -2
[     1] ---- shim_stat("/lib/x86_64",0x71102b0dd120) = -2
[     1] ---- shim_openat(AT_FDCWD,"/lib/libc.so.6",O_RDONLY|2000000,0000) = 3
[     1] ---- shim_read(3,0x71102b0dd288,832) ... read
[     1] bkeep_unmapped: 0x7fb057d98000-0x7fb057d9c000 => filebuf
[     1] ---- return from shim_read(...) = 832
[     1] ---- shim_fstat(3,0x71102b0dd120) = 0
[     1] ---- shim_mmap(0x0,8192,PROT_READ|PROT_WRITE,MAP_PRIVATE|MAP_ANON,-1,0) ... mmap
[     1] bkeep_unmapped: 0x711028ab4000-0x711028ab6000
[     1] heap top adjusted to 0x711028ab4000
[     1] ---- return from shim_mmap(...) = 0x711028ab4000
[     1] ---- shim_mmap(0x0,3869368,PROT_READ|PROT_EXEC,MAP_PRIVATE|0x800,3,0) ... mmap
[     1] bkeep_unmapped: 0x711028703000-0x711028ab4000
[     1] heap top adjusted to 0x711028703000
[     1] ---- return from shim_mmap(...) = 0x711028703000
[     1] ---- shim_mprotect(0x7110288aa000,2097152,PROT_NONE) ... mprotect
[     1] bkeep_mprotect: 0x7110288aa000-0x711028aaa000
[     1] ---- return from shim_mprotect(...) = 0
[     1] ---- shim_mmap(0x711028aaa000,24576,PROT_READ|PROT_WRITE,MAP_PRIVATE|MAP_FIXED|0x800,3,1732608) ... mmap
[     1] mmap: allowing overlapping MAP_FIXED allocation at 0x711028aaa000 with length 24576
[     1] bkeep_mmap: 0x711028aaa000-0x711028ab0000
[     1] ---- return from shim_mmap(...) = 0x711028aaa000
[     1] ---- shim_mmap(0x711028ab0000,15032,PROT_READ|PROT_WRITE,MAP_PRIVATE|MAP_ANON|MAP_FIXED,-1,0) ... mmap
[     1] mmap: allowing overlapping MAP_FIXED allocation at 0x711028ab0000 with length 16384
[     1] bkeep_mmap: 0x711028ab0000-0x711028ab4000
[     1] ---- return from shim_mmap(...) = 0x711028ab0000
[     1] ---- shim_close(3) = 0
[     1] glibc register library /lib/libc.so.6 loaded at 0x711028703000
[     1] add a library for gdb: file:..//../../../Runtime/libc.so.6
[     1] set fs_base to 0x711028ab5500
[     1] ---- shim_arch_prctl(4098,0x711028ab5500) = 0x00000000
[     1] ---- shim_mprotect(0x711028aaa000,16384,PROT_READ) ... mprotect
[     1] bkeep_mprotect: 0x711028aaa000-0x711028aae000
[     1] ---- return from shim_mprotect(...) = 0
[     1] ---- shim_mprotect(0x7fb057ff5000,4096,PROT_READ) ... mprotect
[     1] bkeep_mprotect: 0x7fb057ff5000-0x7fb057ff6000
[     1] ---- return from shim_mprotect(...) = 0
[     1] ---- shim_mprotect(0x711028cdb000,4096,PROT_READ) ... mprotect
[     1] bkeep_mprotect: 0x711028cdb000-0x711028cdc000
[     1] ---- return from shim_mprotect(...) = 0
[     1] ---- shim_openat(AT_FDCWD,"/dev/urandom",O_RDONLY|4400,0000) = 3
[     1] ---- shim_poll(0x71102b0dda08,1,-1) ... poll
*** An unexpected memory fault occurred inside PAL. Exiting the thread. (PID = 17230, TID = 17230, RIP = +0x00003bbd) ***
^Z
[1]+  Stopped                 ./pal_loader poll
```

**Comment by dimakuv** [comment]
@mkow is correct. Yes, there was a bug regarding polling special files like `/dev/urandom` and this bug was fixed in PR #1253.

@shaoyuta You are using a rather old Graphene version. Please update to the latest master -- it contains the fix in PR #1253. And try your toy examples and your Node.JS program again.

**Comment by shaoyuta** [comment]
@dimakuv, @mkow 
I updated graphene to latest master ( 8ca30edbfaeb2fadc9aa43f2290fe24d101bbe94 ), Yes, all thing is OK now. Thanks a lot
